### PR TITLE
Paluch-Elbassoni-Zuylen MAX ATSP approximation

### PR DESCRIPTION
**Pull request by kpotepa**
Implementation of:
- 2/3-approximation for MAX ATSP ([paper](https://drops.dagstuhl.de/opus/volltexte/2012/3412/pdf/28.pdf))
- 5/2-approximation for shortest common superstring via reduction to MAX ATSP

Description will be added soon.

### CODE REVIEW

**Comment by v3ctor** [comment]
sussy

File: shortest_common_superstring/paluch_elbassoni_zuylen.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  cover = set()
+  for i, j in itertools.product(range(n), repeat = 2):
+    if i != j and _is_half_edge_matched(matching, i, j):
+      cover.add((i, j))
+  return cover
+
+def _get_cycles(cover):
+  """Iterate over weakly connected components of the cover,
+  which are cycles if we ignore the edge direction."""
+  neighbours, visited = {}, set()
```
Comment (krzysztof-turowski) [inline shortest_common_superstring/paluch_elbassoni_zuylen.py:65]: Consider using for neighbours:
- import collections
- neighbours = collections.defaultdict(set)
This way you do not need L67-68 I guess

File: shortest_common_superstring/paluch_elbassoni_zuylen.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    A.remove(edge_a)
+    B.remove(edge_b)
+    C += [edge_a, edge_b]
+  elif len(A) == len(cycle):
+    # Cycle contains single directed path. Move one edge from A to C.
+    undirected_edge = next(e for e in A if e not in C)
+    A.remove(undirected_edge)
+    C.append(undirected_edge)
+
+  return (A, B, C)
```
Comment (krzysztof-turowski) [inline shortest_common_superstring/paluch_elbassoni_zuylen.py:115]: You can omit parentheses

File: shortest_common_superstring/paluch_elbassoni_zuylen.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+def _concatenate_path_set(edges, vertex_count):
+  """Given a set of edges that form a node-disjoint path set,
+  build ATSP path that contains all of them."""
+  successors = dict(edges)
+  vertices_with_predecessors = set(v for u, v in edges)
```
Comment (krzysztof-turowski) [inline shortest_common_superstring/paluch_elbassoni_zuylen.py:121]: set(v for _, v in edges) so that linter won't scream errors

**Review by krzysztof-turowski — COMMENTED** [review]
Looks promising, waiting for the text to appear

**Comment by kpotepa** [comment]
Thank you for comments! I addressed the feedback and added text.

**Review by krzysztof-turowski — APPROVED** [review]
Thank you very much. Very clear description, very readable code, I am impressed.

**Pull request merged**